### PR TITLE
fix(split-factories): some fixes

### PR DESCRIPTION
**Pull request by nikophil**
PHPStan fixes:
- using `$this` inside factory was problematic with phpstan
- using `self::findOrCreate()` or any static method was also problematic
- phpstan did not understood factoryCollection->create() would return a list<>
- auto-detect for a `FactoryCollection` whether if the target class has persistence or not 

bug fixes:
- factory parameters are not invokable
- we must apply "before instantiate" callbacks before normalizing step
- `@ORM\Entity` and `@ODM\Document` annotations should be checked along with attributes to check if a class is persistable

### CODE REVIEW

File: src/BaseFactory.php [hunk 1]
```diff
@@ -51,8 +51,9 @@ public function __call(string $name, array $arguments): array
 
     /**
      * @param Attributes|string $attributes
+     * @final
      */
-    final public static function new(array|callable|string $attributes = [], string ...$states): static
+    public static function new(array|callable|string $attributes = [], string ...$states): static
     {
```
Comment (nikophil) [inline src/BaseFactory.php:57]: this change is made to deprecate the usage of `PersistentObjectFactory` with "non-persistable" classes

File: tests/Fixtures/Factories/PostFactory.php [hunk 2]
```diff
@@ -2,13 +2,33 @@
...
+ * @method static Post|Proxy                     find(object|array|mixed $criteria)
+ * @method static Post|Proxy                     findOrCreate(array $attributes)
+ * @method static Post|Proxy                     first(string $sortedField = 'id')
+ * @method static Post|Proxy                     last(string $sortedField = 'id')
+ * @method static Post|Proxy                     random(array $attributes = [])
+ * @method static Post|Proxy                     randomOrCreate(array $attributes = []))
+ * @method static Post[]|Proxy[]                 all()
+ * @method static Post[]|Proxy[]                 findBy(array $attributes)
+ * @method static Post[]|Proxy[]                 randomSet(int $number, array $attributes = []))
+ * @method static Post[]|Proxy[]                 randomRange(int $min, int $max, array $attributes = []))
```
Comment (nikophil) [inline tests/Fixtures/Factories/PostFactory.php:29]: I added those, because some were creating problems with phpstan, now I'm also analyzing this factory with `make sca`

File: tests/allowed-deprecations.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+[
+    {
+        "location": "Zenstruck\\Foundry\\Persistence\\PersistentObjectFactory",
+        "message": "The \"Zenstruck\\Foundry\\BaseFactory::new()\" method is considered final. It may change without further notice as of its next major version. You should not extend it from \"Zenstruck\\Foundry\\Persistence\\PersistentObjectFactory\".",
```
Comment (nikophil) [inline tests/allowed-deprecations.json]: this deprecation occurs in all tests, that's why I've added to the "allowed" list

File: phpunit-dama-doctrine.xml.dist [hunk 4]
```diff
@@ -11,7 +11,11 @@
         <ini name="error_reporting" value="-1"/>
         <env name="KERNEL_CLASS" value="Zenstruck\Foundry\Tests\Fixtures\Kernel"/>
         <env name="USE_DAMA_DOCTRINE_TEST_BUNDLE" value="1"/>
-        <env name="SYMFONY_DEPRECATIONS_HELPER" value="max[self]=0&amp;max[direct]=0&amp;quiet[]=indirect&amp;quiet[]=other"/>
+        <!--
+            max[self]=1 represents "The Zenstruck\Foundry\BaseFactory::new() method is considered final" deprecation.
+            It could be removed in 2.0
+        -->
+        <env name="SYMFONY_DEPRECATIONS_HELPER" value="max[self]=1&amp;max[direct]=0&amp;quiet[]=indirect&amp;quiet[]=other"/>
```
Comment (nikophil) [inline phpunit-dama-doctrine.xml.dist:18]: I tried to use the baseline or ignore files features of the bridge, but it still fails with sf 5.4

File: src/Object/ObjectFactory.php [hunk 5]
```diff
@@ -128,4 +120,24 @@ final protected function instantiate(array $parameters): array
...
     }
+
+    /**
+     * @param Attributes $attributes
+     *
+     * @internal
+     */
+    final protected function mergedAttributes(array|callable $attributes): array
+    {
+        $attributes = parent::mergedAttributes($attributes);
```
Comment (nikophil) [inline src/Object/ObjectFactory.php:131]: I'm not found of this kind of thing...

I'm afraid we'd enter in an "inheritance hell" at some point (`UserlandFactory` extends `PersistenceObjectFactory` extends `ObjectFactory` extends `BaseFactory`)
I'm really wondering if we cannot find another design where we'd use composition over inheritance, but I cannot find a nice pattern :shrug:

File: src/Object/ObjectFactory.php [hunk 6]
```diff
@@ -112,14 +112,6 @@ final protected function instantiate(array $parameters): array
         // execute "lazy" values
         $parameters = LazyValue::normalizeArray($parameters);
 
-        foreach ($this->beforeInstantiate as $hook) {
-            $parameters = $hook($parameters, static::class());
-
-            if (!\is_array($parameters)) {
-                throw new \LogicException('Before Instantiate event callback must return an array.');
-            }
-        }
```
Comment (nikophil) [inline src/Object/ObjectFactory.php:121]: we must apply before instantiate callbacks before normalization occurs

File: src/BaseFactory.php [hunk 7]
```diff
@@ -265,9 +262,10 @@ protected function normalizeAttribute(mixed $value, string $name): mixed
     }
 
     /**
+     * @internal
      * @param Attributes $attributes
      */
-    final protected function mergedAttributes(array|callable $attributes): array
+    protected function mergedAttributes(array|callable $attributes): array
```
Comment (nikophil) [inline src/BaseFactory.php:268]: this change is used to override `mergedAttributes` in `ObjectFactory`, in order to apply "before instantiate" callbacks

File: src/Persistence/PersistenceManager.php [hunk 8]
```diff
@@ -117,6 +122,29 @@ public function isPersistEnabled(): bool
...
+        if (!$attributes) {
+            $attributes = self::annotationReader()?->getClassAnnotation($reflectionClass, ORMEntityAttribute::class);
+        }
+
+        if (!$attributes) {
+            $attributes = self::annotationReader()?->getClassAnnotation($reflectionClass, ODMDocumentAttribute::class);
+        }
+
+        return (bool) $attributes;
+    }
```
Comment (nikophil) [inline src/Persistence/PersistenceManager.php:146]: I know a better way to do this would have been to check doctrine's mapping, and check whether type is set as attribute or annotation, but this way I can check it statically

File: src/Persistence/PersistentObjectFactory.php [hunk 9]
```diff
@@ -301,10 +310,6 @@ final protected function normalizeAttribute(mixed $value, string $name): mixed
             return $value->isPersisted() ? $value->refresh()->object() : $value->object();
         }
 
-        if (\is_callable($value)) {
-            $value = $value();
-        }
```
Comment (nikophil) [inline src/Persistence/PersistentObjectFactory.php:306]: This was coming from a misunderstanding from my side... I've always thought we were allowed to do something like this:
```php
class SomeFactory extends ModelFactory
{
    protected function getDefaults(): array
    {
		return [
			'field' => fn() => FieldFacotry::findOrCreate()
        ];
	}
}
```
but it actually introduce a bug if the value passed is invokable (I had the case in one of the project where I tested this version)

File: tests/Fixtures/Factories/PostFactory.php [hunk 10]
```diff
@@ -22,6 +42,24 @@ public static function class(): string
...
+    }
+
+    /**
+     * @return FactoryCollection<Proxy<Post>>
+     * This method is here to test an edge case with the phpstan extension
+     */
+    public function methodUsingThis(): FactoryCollection
+    {
+        return $this->sequence([]);
+    }
```
Comment (nikophil) [inline tests/Fixtures/Factories/PostFactory.php:61]: this was easier to put these method here but I can create a factory specific for static analysis

**Review by kbond — APPROVED** [review]
Wow, nice work here!
